v8 RPC API (#134)
MetaMask/metamask-docs: docs/guide/rpc-api.md

# The Ethereum RPC API
# RPC API

MetaMask uses the `ethereum.sendAsync()` (and soon, `ethereum.request()`) API to wrap an RPC API which is based on an interface exposed by all Ethereum clients, with some extra methods that are provided by MetaMask, as a key-holding signer. You can look up how to pass these methods to the `window.ethereum` object [here](./Ethereum-Provider).
MetaMask uses the [`ethereum.request(args)` method](./ethereum-provider.html#ethereum-request-args) to wrap an RPC API.

> For the full API, please see [EIP 1474](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1474.md).

Reviewer: Now that the title of this page has changed, it no longer seems appropriate to call this the "full" API. The Ethereum wiki link below documents a wider variety of RPC methods than this EIP does, and even that doesn't show them all (e.g. `eth_subscribe`)

Reviewer: Maybe we can link nothing on this line, and link both below? 🤔 Not sure.

All RPC method requests can return errors.
Make sure to handle errors for every call to `ethereum.request(args)`.
:::

## Table of Contents

[[toc]]

## Ethereum JSON-RPC Methods

For the Ethereum JSON-RPC API, please see [the Ethereum wiki](https://eth.wiki/json-rpc/API#json-rpc-methods).

Important methods from this API include:

- [`eth_accounts`](https://eth.wiki/json-rpc/API#eth_accounts)
- [`eth_call`](https://eth.wiki/json-rpc/API#eth_call)
- [`eth_getBalance`](https://eth.wiki/json-rpc/API#eth_getBalance)
- [`eth_sendTransaction`](https://eth.wiki/json-rpc/API#eth_sendTransaction)
- [`eth_sign`](https://eth.wiki/json-rpc/API#eth_sign)

## Permissions

These RPC methods are called in order to access the user's accounts.

> In the future, they will also be used to request access to other permissioned MetaMask functionality.

Author: 2862988

If you're interested in learning more about the theory behind this _capability_-inspired permissions system, we encourage you to take a look at [EIP-2255](https://eips.ethereum.org/EIPS/eip-2255).

### eth_requestAccounts

::: tip EIP-1102
This method is specified by [EIP-1102](https://eips.ethereum.org/EIPS/eip-1102).
It is equivalent to the deprecated [`ethereum.enable()`](./ethereum-provider.html#ethereum-enable) provider API method.

Under the hood, it calls [`wallet_requestPermissions`](#wallet-requestpermissions) for the `eth_accounts` permission.
Since `eth_accounts` is currently the only permission, this method is all you need for now.
:::

```typescript
ethereum.request({ method: 'eth_requestAccounts' }): Promise<string[]>
```

Requests that the user provides an Ethereum address to be identified by.
Returns a Promise that resolves to an array of a single Ethereum address string.
If the user denies the request, the Promise will reject with a `4001` error.

The request causes a MetaMask popup to appear.
You should only request the user's accounts in response to user action, such as a button click.

> Otherwise, you will popup-spam the user.

Author: 2862988

```javascript
document.getElementById('requestPermissionsButton', requestPermissions);

function requestPermissions() {
  ethereum
    .request({
      method: 'eth_requestAccounts',
      params: [{ eth_accounts: {} }],
    })
    .then((permissions) => {
      const accountsPermission = permissions.find(
        (permission) => permission.parentCapability === 'eth_accounts'
      );
      if (accountsPermission) {
        console.log('eth_accounts permission successfully requested!');
      }
    })
    .catch((error) => {
      if (error.code === 4001) {
        // EIP-1193 userRejectedRequest error
        console.log('Permissions needed to continue.');
      } else {
        console.error(error);
      }
    });
}
```

### wallet_getPermissions

```typescript
ethereum.request({
  method: 'wallet_requestPermissions',
}): Promise<Web3WalletPermission[]>;
```

Gets the caller's current permissions.
Returns a Promise that resolves to an array of `Web3WalletPermission` objects.
If the caller has no permissions, the array will be empty.

## Other RPC Methods

### wallet_registerOnboarding

::: tip
As an API consumer, you are unlikely to have to call this method yourself.

> Please see [`@metamask/onboarding`](https://github.com/MetaMask/metamask-onboarding) for more information.

Reviewer: We should link the docs page about onboarding here instead. The repo README isn't really intended for dapp devs.
```suggestion
Please see the [Onboarding Library documentation](/onboarding-library.html) for more information.
```

:::

```typescript
ethereum.request({ method: 'wallet_registerOnboarding' }): Promise<true>;
```

> Requests that MetaMask completes its onboarding flow for the user.

Reviewer: Whoops; this is not correct.
```suggestion
This method is used to register the requesting site with MetaMask as the initiator of onboarding. In other words, you can use this method to inform MetaMask that you were the one that suggested installing MetaMask. This lets MetaMask redirect the user back to your site after onboarding as completed.
```

Author: 2862988

Returns a Promise that resolves to `true`, or rejects if there's an error.

> This method is intended to be called at the end of the MetaMask onboarding flow.

Reviewer: ```suggestion
This method is intended to be called after MetaMask has been installed, but before the MetaMask onboarding has completed.
```

> The onboarding flow is initiated by our [`@metamask/onboarding`](https://github.com/MetaMask/metamask-onboarding) package, which calls `wallet_registerOnboarding` upon redirecting the user back to the original site.

Reviewer: This sentence seems incorrect. I'm not sure we need to replace it with anything.
```suggestion
```

Instead of calling this method directly, you should use [`@metamask/onboarding`](https://github.com/MetaMask/metamask-onboarding) directly.

### wallet_watchAsset

::: tip EIP-747
This method is specified by [EIP-747](https://eips.ethereum.org/EIPS/eip-747).
:::

```typescript
interface WatchAssetParams {
  type: 'ERC20', // In the future, other standards will be supported
  options: {
    address: string, // The address of the token contract

>     'symbol': string, // A ticker symbol or shorthand, up to 5 characters

Reviewer: Why quote just this property :thinking: 
```suggestion
    symbol: string, // A ticker symbol or shorthand, up to 5 characters
```

Author: Because `symbol` is a reserved keyword, and the syntax highlighting picks it up.

Not sure if it'd be better to quote everything?

Reviewer: Ah, no, this seems fine. I totally forgot it was a reserved keyword.

```javascript
let encryptionPublicKey;

ethereum
  .request({
    method: 'eth_getEncryptionPublicKey',
    params: [accounts[0]], // you must have access to the specified account
  })
  .then((result) => {
    encryptionPublicKey = result;
  })
  .catch((error) => {
    if (error.code === 4001) {
      // EIP-1193 userRejectedRequest error
      console.log('We can encrypt anything without the key.');
    } else {
      console.error(error);
    }
  });
```

#### Encrypting

The point of the encryption key is of course to encrypt things.
Here's an example of how to encrypt a message using [`eth-sig-util`](https://github.com/MetaMask/eth-sig-util):

```javascript
const ethUtil = require('ethereumjs-util');

const encryptedMessage = ethUtil.bufferToHex(
  Buffer.from(
    JSON.stringify(
      sigUtil.encrypt(
        encryptionPublicKey,
        { data: 'Hello world!' },
        'x25519-xsalsa20-poly1305'
      )
    ),
    'utf8'
  )
);
```

### eth_decrypt

```typescript
// these two types are simply hexadecimal strings
type EthereumAddress: string;
type EncryptedMessage: string;
// the decrypted message is the originally encrypted string
type DecryptedMessage: string;

> ethereum.request({

Reviewer 2: I get that the goal of this is to inform the user of the types, but this wouldn't compile because its a mixture of a function invocation with a type signature for that function invocation. You'd need to make ethereum.request a parameterized function eg:

```ts
function request<ParamsType, ReturnType>({ method: string, params: ParamsType }): Promise<ReturnType> {
  // implementation
}

request<[EncryptedMessage, EthereumAddress], DecryptedMessage>({ method: 'eth_decrypt', params: ['0xasdf1a7=', '0xa00a..324r3']
}).then(decryptedMessage => { // do something with decryption });
```

If we have an explainer somewhere that the examples are just for showing types and not real code, then we can just leave it as is maybe?

Author: Hmm, yeah, the goal is mostly to be informative, and I think for most devs, the parameterized function obscures more than it explains, even though it compiles 🤔 

I'm inclined to leave a notice, depending on what we all think. It is a bit ugly to have it listed as TS that's just... syntactically invalid.

Reviewer 2: alternatively...

```ts
// Create an intermediary typesafe function
function ethDecrypt(params: [EncryptedMessage, EthereumAddress]): Promise<DecryptedMessage> {
  return ethereum.request({ method: 'eth_decrypt', params })
}
```

might be cleaner in regards to not relying on Algebraic data types, compiles, etc.

Author: Hm, or, alternatively alternatively, I suppose we do what most JSON-RPC APIs I’ve seen do, e.g. https://kodi.wiki/view/JSON-RPC_API/v10

It’s the same for the Ethereum JSON-RPC API listed on eth.wiki.

Author: @Gudahtt @brad-decker this is the last thing we need to check off for this PR. See: 3dc33db

The addition of `.prettierrc.js` was necessary to create an exception for a particular snippet that used a reserved TypeScript keyword (`symbol`) in an interface.
